### PR TITLE
Enhance Thumbnail Functionality

### PR DESCRIPTION
**Pull request by studybuffalo**
This is a pull request to address the dependency issues caused by conflicting versions of django-newsletter, sorl-thumbnail, and Django. It removes the hard dependency on sorl-thumbnail and provides alternative fallbacks if sorl-thumbnail is not used. With this work it was easy enough to extend functionality to include other thumbnail applications. Thus, I have also included easy-thumbnails support to give users another option and serve as a template for any future support for other thumbnail applications.

# Major Changes
- Removed `sorl-thumbnail` as a dependency.
- Added a new setting (`NEWSLETTER_THUMBNAIL`) that allows user to specify a third party thumbnail application for use with django-newsletter.
- Added support for two thumbnail applications to start: `sorl-thumbnail` and `easy-thumbnails`
- Created a template tag that performs rudimentary thumbnailing with Pillow. This acts as a fallback if the user hasn't specified a third party thumbnail application. This has been tested to support common image file formats: JPG, PNG, GIF, BMP, TIFF.
- Reworked the `image` field in the `Article` model to support swapping between thumbnail applications by creating a model field that will inherit from the appropriate application. This avoids issues where Django will try to make a new migration anytime the user swaps the app. By default, this field will fallback to the Django `ImageField` if there is no thumbnail app specified.
- Added logic to the `SubmissionArchiveDetailView` to select an appropriate template to thumbnail any `Article` images. Users can override the `newsletter/message/thumbnail/django_newletter.html` or `newsletter/message/message.html` template if they need to implement a different template.
- Added logic to the admin `ArticleInline` class to make use of the appropriate thumbnail application widget as specified by the user (or the normal Django widet if not specified).
- Added unit tests for all new functionality
- Updated relevant documentation sections to outline the new functionality and modifications to the install steps.
- Removed testing steps from `travis.yml` that stops installing directly from `requirements.txt`. This forces the tests to install as though it were a proper PyPI package and catch any odd requirement issues.

# Other Comments
- the `upgrading.rst` file has not been updated yet, as I am not sure which version this would fall under and if there are major changes to still take place.

# Questions before review
- I am not sure if we want to go ahead with the basic thumbnail functionality via Pillow as a fallback or not. I added it since the work was already done and it was minimal effort at this point. That said, I understand if we don't want this application to dive into something that could become quite complex. I think this works for the common uses, but we can't guarantee that for anything unusual. If we opt not to use it, I think the fallback becomes to just display the image as is and resize it via HTML attributes (i.e. no actual performance benefits, just making sure it visually looks the same as the other thumbnail applications).
- I opted to create a more rigid implementation of specifying the thumbnail application. We could go with a situation where the user can specify things down to the specific model field they want to use, the admin mixin, the templates, etc. I felt that it wasn't functionality most users were going to use and we should encourage developers to add their thumbnail application of choice as an option within django-newsletter.

Relevant Issues: #302, #303

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -39,8 +39,6 @@ install:
   # Latest PIP uses wheel by default
   - pip install --upgrade pip
   - pip install $DJANGO
-  - pip install -r requirements.txt
-  - pip install -r requirements_test.txt
```
Comment (dokterbob) [inline .travis.yml:43]: Why are these removed?

File: newsletter/admin.py [hunk 2]
```diff
@@ -228,11 +245,22 @@ class ArticleInline(AdminImageMixin, StackedInline):
...
+
     if newsletter_settings.RICHTEXT_WIDGET:
-        formfield_overrides = {
+        _formfield_overrides = {
             models.TextField: {'widget': newsletter_settings.RICHTEXT_WIDGET},
         }
 
+    # https://easy-thumbnails.readthedocs.io/en/latest/usage/#forms
+    if newsletter_settings.THUMBNAIL == 'easy-thumbnails':
+        _formfield_overrides = {
```
Comment (dokterbob) [inline newsletter/admin.py:258]: This *seems* to override the `_formfield_overrides` when a RichtextWidget is set.

It's also not clear to me why a separate `_formield_overrides` is used instead of directly setting/updating `formfield_overrides`.

File: newsletter/templatetags/newsletter_thumbnails.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    """Takes the provided ImageFile and returns a thumbnail URL."""
+    # Split the tag into its components
+    # Second item = image argument; last item = context variable
+    tag_args = token.split_contents()
+
+    # Retrieve the source image argument as a filter expression
+    source_image = parser.compile_filter(tag_args[1])
+
+    # Return the thumbnail
+    return ThumbnailNode(source_image, tag_args[-1])
```
Comment (dokterbob) [inline newsletter/templatetags/newsletter_thumbnails.py:111]: I do think providing (and thus maintaining!) thumbnail code is really out of scope from django-newsletter!

My suggestion would be:
1. For now, to default to `sorl-thumbnail` for backwards-compatibility but to yield a `DeprecationError` if no thumbnail facility is explicitly configured.
2. In the next major version, failing hard if none is defined. We might make sure there's a no-op thumbnail facility by default.

I am open to your suggestions on this one though!

Super happy with this great contrib!

File: .travis.yml (hunk 1 above)
Comment (studybuffalo) [reply .travis.yml:43]: I believe this was because I was noticing some odd behaviour in testing. I unfortunately cannot recall exactly what was happening, but I believe I was getting tests that passed locally with the `setup.py test` command, but then failing in Travis-CI. Removing these lines resolved the issue and all other tests continue to pass. My Travis-CI history unfortunately has since removed the offending tests from the history.

If I were to make an educated guess, I think we were getting weird version conflicts because we install these packages into the environment and then install things again with the `setup.py test` command.

If we need a definitive answer I can add the lines back and see what happens in testing!

File: newsletter/admin.py (hunk 2 above)
Comment (studybuffalo) [reply newsletter/admin.py:258]: That is correct. Ultimately this is largely a personal stylistic choice so that we only have one spot where we set the attribute value, but we could achieve the same effect with if/elif/else statements and set the value directly.

File: newsletter/templatetags/newsletter_thumbnails.py (hunk 3 above)
Comment (studybuffalo) [reply newsletter/templatetags/newsletter_thumbnails.py:111]: I like that plan and I think it makes perfect sense. I have no doubts there are countless pitfalls with proper thumbnail generation we aren't accounting for here and it could create some concerning feature creep.

I'll look at making the suggested updates and remove this block (and updating the other impacted code sections).

**Comment by studybuffalo** [comment]
Thanks for taking the time to review and provide the feedback @dokterbob - really appreciate it!

I will try and find some time to make the changes around the thumbnailer functionality in the near future and get another version out.

**Comment by studybuffalo** [comment]
Okay, had some time and pulled out the thumbnail functionality. I left comments to some of your other questions and can make any additional changes as needed quite quickly. Thanks!

File: newsletter/templatetags/newsletter_thumbnails.py (hunk 3 above)
Comment (studybuffalo) [reply newsletter/templatetags/newsletter_thumbnails.py]: Set this up to raise a `DeprecationWarning` if no `NEWSLETTER_THUMBNAIL` is declared and then default to `sorl-thumbnail`. I have set the warning to say this will become a mandatory declaration for version 0.11.0 (I assume the warning itself will kick in for version 0.10.0). I updated some details in the installation and setting docs to cover this change as well.

File: .travis.yml (hunk 1 above)
Comment (dokterbob) [reply .travis.yml:43]: I would prefer you take this file straight from master, where CI seems working just fine. :)

File: newsletter/admin.py (hunk 2 above)
Comment (dokterbob) [reply newsletter/admin.py:258]: I would prefer setting them directly as I feel non-functional code does not contribute to clarity.

Comment (studybuffalo) [reply newsletter/admin.py]: Updated with requested changes.

File: .travis.yml (hunk 1 above)
Comment (studybuffalo) [reply .travis.yml]: Updated with requested changes.

**Comment by dokterbob** [comment]
Merged, at last! :D

Thanks, @studybuffalo!

Ref: https://github.com/dokterbob/django-newsletter/commit/2ce29ba3f00c6edfb87cd5a34fe7f78dd70766e6

**Comment by frennkie** [comment]
@dokterbob the migration that was merged with 2ce29ba is called `0006_switch_from_sorl_imagefield.py`. The "attachment" migration already has the prefix `0006`. I'm sure that is is no technical problem at all. But it might be a bit confusing to developers/users.

What do you think? Is it necessary to rename this one to `0007_switch_from_sorl_imagefield.py` or should we just leave it?

**Comment by dokterbob** [comment]
Oh, that's a silly human bug right there. We might quickly renumber them, before anyone uses it (or else they'll have to manually rewind and re-apply). What do you think?

**Comment by newearthmartin** [comment]
Hey @studybuffalo @dokterbob

As people have pointed out here https://github.com/jazzband/django-newsletter/issues/396 the default installation is breaking because `{% include thumbnail_template %}` complains that the `thumbnail_template` is not found. 

It seems that the `thumbnail_template` context var is set only in `SubmissionArchiveDetailView` and not set in other render methods. For example:

```
    def preview_html(self, request, object_id):
        message = self._getobj(request, object_id)

        if not message.html_template:
            raise Http404(_(
                'No HTML template associated with the newsletter this '
                'message belongs to.'
            ))

        c = {
            'message': message,
            'site': Site.objects.get_current(),
            'newsletter': message.newsletter,
            'date': now(),
            'STATIC_URL': settings.STATIC_URL,
            'MEDIA_URL': settings.MEDIA_URL
        }

        return HttpResponse(message.html_template.render(c))
```
So unless I'm missing something this is a gross bug cause it's never going to find that context variable and it will always raise that exception. It's a shame new users are encountering this exception... since 2020 !!

So I would fix it now, and I'm thinking to add a default value, and I'm wondering what's wrong with adding sorl as a default template for this, since this PR was about removing that dependency. So if you guys can please explain why, so I don't add back what you took away. Thanks!
